### PR TITLE
Enable searching for assemblies in GAC_Arm64 on Windows

### LINKED ISSUE #17814: Import-Module PKI fails with ARM64 powershell core 7.2.5 on ARM64 Windows 11 22H2
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

1. Download ARM64 powershell core 7.2.5 on ARM64 Windows 11 22H2 (22621)
2. Launch pwsh.exe and type "Import-Module PKI"
3. Observe this failure message: "Import-Module: Could not load file or assembly 'Microsoft.CertificateServices.PKIClient.Cmdlets, Culture=neutral, PublicKeyToken=null'. Format of the executable (.exe) or library (.dll) is invalid."

### Expected behavior

```console
No error.
```

### Actual behavior

```console
Import-Module: Could not load file or assembly 'Microsoft.CertificateServices.PKIClient.Cmdlets, Culture=neutral, PublicKeyToken=null'. Format of the executable (.exe) or library (.dll) is invalid.
```

### Error details

```console
Exception             :
    Type           : System.BadImageFormatException
    Message        : Could not load file or assembly 'Microsoft.CertificateServices.PKIClient.Cmdlets, Culture=neutral, PublicKeyToken=null'. Format of the executable
(.exe) or library (.dll) is invalid.
    FileName       : Microsoft.CertificateServices.PKIClient.Cmdlets, Culture=neutral, PublicKeyToken=null
    TargetSite     :
        Name          : LoadBinaryModule
        DeclaringType : Microsoft.PowerShell.Commands.ModuleCmdletBase
        MemberType    : Method
        Module        : System.Management.Automation.dll
    InnerException :
        Type       : System.BadImageFormatException
        Message    : Could not load file or assembly 'C:\WINDOWS\Microsoft.NET\assembly\GAC_64\Microsoft.CertificateServices.PKIClient.Cmdlets\v4.0_10.0.0.0__31bf3856a
d364e35\Microsoft.CertificateServices.PKIClient.Cmdlets.dll'. Format of the executable (.exe) or library (.dll) is invalid.
        FileName   : C:\WINDOWS\Microsoft.NET\assembly\GAC_64\Microsoft.CertificateServices.PKIClient.Cmdlets\v4.0_10.0.0.0__31bf3856ad364e35\Microsoft.Certificate
Services.PKIClient.Cmdlets.dll
        TargetSite :
            Name          : LoadFromPath
            DeclaringType : System.Runtime.Loader.AssemblyLoadContext
            MemberType    : Method
            Module        : System.Private.CoreLib.dll
        Source     : System.Private.CoreLib
        HResult    : -2147024703
        StackTrace :
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Management.Automation.PowerShellAssemblyLoadContext.Resolve(AssemblyLoadContext loadContext, AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.GetFirstResolvedAssemblyFromResolvingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingResolvingEvent(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
    Source         : System.Management.Automation
    HResult        : -2147024703
    StackTrace     :
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadBinaryModule(PSModuleInfo parentModule, String moduleName, String fileName, Assembly assemblyToLoad, String
moduleBase, SessionState ss, ImportModuleOptions options, ManifestProcessingFlags manifestProcessingFlags, String prefix, Boolean loadTypes, Boolean loadFormats,
Boolean& found, String shortModuleName, Boolean disableFormatUpdates)
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadModuleNamedInManifest(PSModuleInfo parentModule, ModuleSpecification moduleSpecification, String moduleBase,
Boolean searchModulePath, String prefix, SessionState ss, ImportModuleOptions options, ManifestProcessingFlags manifestProcessingFlags, Boolean loadTypes, Boolean
loadFormats, Object privateData, Boolean& found, String shortModuleName, Nullable`1 manifestLanguageMode)
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadModuleManifest(String moduleManifestPath, ExternalScriptInfo manifestScriptInfo, Hashtable data, Hashtable
localizedData, ManifestProcessingFlags manifestProcessingFlags, Version minimumVersion, Version maximumVersion, Version requiredVersion, Nullable`1
requiredModuleGuid, ImportModuleOptions& options, Boolean& containedErrors)
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadModule(PSModuleInfo parentModule, String fileName, String moduleBase, String prefix, SessionState ss, Object
privateData, ImportModuleOptions& options, ManifestProcessingFlags manifestProcessingFlags, Boolean& found, Boolean& moduleFileFound)
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadUsingExtensions(PSModuleInfo parentModule, String moduleName, String fileBaseName, String extension, String
moduleBase, String prefix, SessionState ss, ImportModuleOptions options, ManifestProcessingFlags manifestProcessingFlags, Boolean& found, Boolean& moduleFileFound)
   at Microsoft.PowerShell.Commands.ModuleCmdletBase.LoadUsingModulePath(PSModuleInfo parentModule, Boolean found, IEnumerable`1 modulePath, String name, SessionState
ss, ImportModuleOptions options, ManifestProcessingFlags manifestProcessingFlags, PSModuleInfo& module)
   at Microsoft.PowerShell.Commands.ImportModuleCommand.ImportModule_LocallyViaName(ImportModuleOptions importModuleOptions, String name)
   at Microsoft.PowerShell.Commands.ImportModuleCommand.ImportModule_LocallyViaName_WithTelemetry(ImportModuleOptions importModuleOptions, String name)
   at Microsoft.PowerShell.Commands.ImportModuleCommand.ProcessRecord()
   at System.Management.Automation.Cmdlet.DoProcessRecord()
   at System.Management.Automation.CommandProcessor.ProcessRecord()
CategoryInfo          : NotSpecified: (:) [Import-Module], BadImageFormatException
FullyQualifiedErrorId : System.BadImageFormatException,Microsoft.PowerShell.Commands.ImportModuleCommand
InvocationInfo        :
    MyCommand        : Import-Module
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 5
    Line             : Import-Module PKI
    PositionMessage  : At line:1 char:1
                       + Import-Module PKI
                       + ~~~~~~~~~~~~~~~~~
    InvocationName   : Import-Module
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.2.5
PSEdition                      Core
GitCommitId                    7.2.5
OS                             Microsoft Windows 10.0.25173
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

The code to enable searching the GAC was limited to GAC_32 and GAC_64 which fails on Windows Arm64 systems.  Fix is to use GAC_Arm64 if pwsh is running as a Arm64 process.

Manually validated with PKI module with (requires admin):

```powershell
New-SelfSignedCertificate -DnsName "www.fabrikam.com", "www.contoso.com" -CertStoreLocation "cert:\LocalMachine\My"
```

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/17814

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by mcfi — APPROVED** [review]

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 1]
```diff
@@ -371,8 +371,14 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
...
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                        if (RuntimeInformation.ProcessArchitecture == Architecture.Arm64)
+                        {
+                            _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_Arm64";
+                        }
+                        else
+                        {
+                            _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                        }
```
Comment (iSazonov) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:381]: Interpolated string uses implicitly CurrentCulture but our convention is (1) use InvariantCulture in Engine, (2) use culture _explicitly_.

```suggestion
                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64": "GAC_64";
                         _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}";
```

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 2]
```diff
@@ -371,8 +371,8 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
                 {
                     if (string.IsNullOrEmpty(_gacPath64))
                     {
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64": "GAC_64";
+                         _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}";
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:375]:
```suggestion
                         gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 3]
```diff
@@ -371,8 +371,8 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
                 {
                     if (string.IsNullOrEmpty(_gacPath64))
                     {
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64": "GAC_64";
+                         gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:375]:
```suggestion
                         _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 4]
```diff
@@ -371,8 +371,8 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
                 {
                     if (string.IsNullOrEmpty(_gacPath64))
                     {
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64": "GAC_64";
+                         _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:375]:
```suggestion
                        _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 5]
```diff
@@ -371,8 +371,8 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
                 {
                     if (string.IsNullOrEmpty(_gacPath64))
                     {
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64": "GAC_64";
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:374]:
```suggestion
                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64" : "GAC_64";
```

File: src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs [hunk 6]
```diff
@@ -371,8 +371,8 @@ private bool TryFindInGAC(AssemblyName assemblyName, out string assemblyFilePath
                 {
                     if (string.IsNullOrEmpty(_gacPath64))
                     {
-                        // cache value of '_gacPath64' folder in member variable.
-                        _gacPath64 = $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}GAC_64";
+                       var gacName = RuntimeInformation.ProcessArchitecture == Architecture.Arm64 ? "GAC_Arm64" : "GAC_64";
+                        _gacPath64 = string.Create(CultureInfo.InvariantCulture, $"{_winDir}{dirSeparator}Microsoft.NET{dirSeparator}assembly{dirSeparator}{gacName}");
```
Comment (daxian-dbw) [inline src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:375]: Can we use `Path.Join(string, string, string, string)` (or `Path.Combine(s, s, s, s)`) to construct the path?
If not too much trouble, maybe also replace other `"${...}{dirSeparator}{...}..."` instances with `Path.Join` (or `Path.Combine`).

Comment (SteveL-MSFT) [reply src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs:375]: Sure, was just using the current style, but makes sense

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Review by iSazonov — APPROVED** [review]

**Pull request merged**